rosetta/converter/converter.go

```diff
@@ -0,0 +1,120 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package converter
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/onflow/cadence"
+	"github.com/onflow/cadence/encoding/json"
+	"github.com/onflow/flow-go/model/flow"
+
+	"github.com/optakt/flow-dps/models/dps"
+	"github.com/optakt/flow-dps/rosetta/identifier"
+	"github.com/optakt/flow-dps/rosetta/object"
+)
+
+type Converter struct {
+	deposit    flow.EventType
+	withdrawal flow.EventType
+}
+
+func New(gen Generator) (*Converter, error) {
+	deposit, err := gen.TokensDeposited(dps.FlowSymbol)
+	if err != nil {
+		return nil, fmt.Errorf("unable to generate deposit event type: %w", err)
+	}
+	withdrawal, err := gen.TokensWithdrawn(dps.FlowSymbol)
+	if err != nil {
+		return nil, fmt.Errorf("unable to generate withdrawal event type: %w", err)
+	}
+
+	c := Converter{
+		deposit:    flow.EventType(deposit),
+		withdrawal: flow.EventType(withdrawal),
+	}
+
+	return &c, err
+}
+
+func (c *Converter) EventToOperation(event flow.Event) (operation *object.Operation, isRelevant bool, err error) {
+	switch event.Type {
+	case c.deposit, c.withdrawal:
+		isRelevant = true // Supported types, event is relevant for us.
+	default:
+		return nil, false, nil // This type of event is irrelevant for us.
+	}
+
+	// Decode the event payload into a Cadence value and cast it to a Cadence event.
+	value, err := json.Decode(event.Payload)
+	if err != nil {
+		return nil, isRelevant, fmt.Errorf("could not decode event: %w", err)
+	}
+	e, ok := value.(cadence.Event)
+	if !ok {
+		return nil, isRelevant, fmt.Errorf("could not cast event: %w", err)
+	}
+
+	// Ensure that there are the correct amount of fields.
+	if len(e.Fields) != 2 {
+		return nil, isRelevant, fmt.Errorf("invalid number of fields (want: %d, have: %d)", 2, len(e.Fields))
+	}
+
+	// The first field is always the amount and the second one the address.
+	// The types coming from Cadence are not native Flow types, so primitive types
+	// are needed before they can be converted into proper Flow types.
+	vAmount := e.Fields[0].ToGoValue()
+	uAmount, ok := vAmount.(uint64)
+	if !ok {
+		return nil, isRelevant, fmt.Errorf("could not cast amount (%T)", vAmount)
+	}
+	vAddress := e.Fields[1].ToGoValue()
+	bAddress, ok := vAddress.([flow.AddressLength]byte)
+	if !ok {
+		return nil, isRelevant, fmt.Errorf("could not cast address (%T)", vAddress)
+	}
+
+	// Convert the amount to a signed integer that it can be inverted.
+	amount := int64(uAmount)
+	// Convert the address bytes into a native Flow address.
+	address := flow.Address(bAddress)
+
+	// In the case of a withdrawal, invert the amount value.
+	if event.Type == c.withdrawal {
+		amount = -amount
+	}
```

Comment:
This is fine for now as well. However, the code suggestion I wrote was based on how this code will evolve in the future, and having the switch statement makes a lot more sense in that context. We can extract all common values in common code, and then use a switch in the end to only set the fields that are different between event types. I think that flow is clearer.

```go
switch event.Type {
  case deposit:
    op.Type = dps.OperationTransfer
    op.Amount = amount
  case withdrawal:
    op.Type = dps.OperationTransfer
    op.Amount = -amount
  case stake:
    op.Type = dps.OperationStake
    op.Amount = amount
  case unstake:
    op.Type = dps.OperationStake
    op.Amount = -amount
  // same for delegate & lock
  default:
    return nil, false, fmt.Errorf("invalid event type (%T)", event)
}
```